### ISSUE TITLE
[BUG] Strange symbols instead of colored text

### ISSUE THREAD

**Issue by xairaven** [gyscos/cursive#800]
**Describe the bug**
Hi! I am writing my own version of the project [github.com/kognise/arpchat](https://github.com/kognise/arpchat) for educational purposes. I encountered an issue where strange symbols appear next to the colored text in my project. By colored text, I mean text styled using methods from `crossterm::style::stylize::Stylize`, for example, `"some text".dark_grey().to_string()`. You can see what I mean in the screenshot. I will provide some code snippets for reproduction, but it would be easier to clone the project I mentioned above and run it. <br>
An interesting detail I noticed is that if I set the **Cursive** version to `0.20.0`, everything works as expected. But starting from version `0.21.0`, it breaks. I can’t simply use version `0.20.0` because it only works with that original project, where I tested the transitions between versions (it works on nightly old Rust). My project is written using the stable Rust version 1.81, and for some reason, it doesn’t compile with **Cursive** versions earlier than `0.21.1`. <br>
I apologize if I missed something, and this is my mistake and not a bug. I’d be happy to collaborate.

**To Reproduce**
```rust
use cursive::backends::crossterm::crossterm::style::Stylize;
use cursive::views::{LinearLayout, NamedView, TextView};
use cursive::Cursive;
ui::cursive_extension::append_txt(
                siv,
                "chat_area",
                "some_text".dark_grey().to_string(),
            );
```

Another file:
```rust
use cursive::traits::Nameable;
use cursive::utils::markup::StyledString;
use cursive::views::{LinearLayout, TextView};
use cursive::Cursive;

/// Append a new `TextView` to the `LinearLayout` with the provided parent name.
pub fn append_txt<S>(siv: &mut Cursive, parent_id: &str, content: S)
where
    S: Into<StyledString>,
{
    siv.call_on_name(parent_id, |parent: &mut LinearLayout| {
        parent.add_child(TextView::new(content));
    });
}
```

**Expected behavior**
Colored text instead of strange symbols.

**Screenshots**
Cursive version 0.21 (changed just the version in `Cargo.toml`):
![image](https://github.com/user-attachments/assets/12ff450c-89b7-422c-8f68-af1e8df4e975)

Cursive version 0.20:
![image](https://github.com/user-attachments/assets/eecdc86f-9358-4ba4-a4cd-ab659a33b86a)

**Environment**
* OS: Windows 10 22H2
* Backend used: I don't know. I create cursive object with `cursive::default()`. If I use use feature `crossterm-backend`, result stays the same.
* Current locale (run `locale` in a terminal): `System Locale:  en-us;English (United States); Input Locale: en-us;English (United States)`
* Cursive version (from crates.io, from git, ...): 0.21.1

**Comment by xairaven** [comment]
Update: I just managed to compile my project using this line in Cargo.toml: 
```toml
cursive = { version = "0.20.0", default-features = false, features = ["toml", "crossterm-backend"]}
```

Yes, as expected, everything works on version 0.20. However, if I use version `0.21.1` like this:
```toml
cursive = { version = "0.21.1", default-features = false, features = ["toml", "crossterm-backend"]}
```
Or this:
```toml
cursive = { version = "0.21.1", features = ["toml"]}
```
I see strange characters instead of colors, meaning the situation is the same to what I described earlier. You can check the screenshots to understand what I mean.

**Comment by gyscos** [comment]
Hi, and thanks for the report!

The main issue is that `"some_text".dark_grey().to_string()` returns some text with "ansi" markup, which cursive doesn't understand by default. Cursive 0.21 brought some performance optimization (an output buffer cell grid) which makes this even more visible - but even with 0.20 you would run into various issues.

Note that cursive does support styled text with [`StyledString`](https://docs.rs/cursive/latest/cursive/utils/markup/type.StyledString.html), just not a regular `String` with ansi codes inside.

You _can_ parse ansi text: https://docs.rs/cursive/latest/cursive/utils/markup/ansi/fn.parse.html
But it's a bit roundabout to use crossterm to generate an ansi string, just to parse it back.

You could instead directly build the cursive-native `StyledString`, for example with [`StyledString::single_span`](https://docs.rs/cursive/latest/cursive/utils/markup/type.StyledString.html#method.single_span) if you only need a single style for the entire string. 

There is also `cursup`, a simple markup language: https://docs.rs/cursive/latest/cursive/utils/markup/cursup/index.html

**Comment by xairaven** [comment]
Thank you so much for your response! I used the `ansi::parse()` function, and everything works. Sorry for bothering you, I couldn't find the information myself, although there were similar examples in the documentation🤦‍♂️ I think the issue is resolved, so it can be closed. However, I also created a pull request with an example of my problem. If you find it relevant, I'd be happy to help someone else who might encounter the same issue when upgrading to the new version and misses the `ansi` example. My example directly addresses the issue I faced: https://github.com/gyscos/cursive/pull/801

**Comment by xairaven** [comment]
@gyscos, By the way, is there a way to remove ANSI codes from a string and get plain text without any styling? For example, I have `"some text".dark_grey().to_string()`. Can I get plain text from this text without ANSI characters? I tried this: 

```rust
let ansi_text = "some text".dark_grey().to_string();
let parsed = SpannedString::<Style>::plain(ansi_text);
format!("{}", parsed.source()).as_str();
```
but it didn’t work. Sorry if this is a silly question.

**Comment by gyscos** [comment]
You have a couple of ways to get that:
* Iterate on the spans, and accumulate the span text content
* Use `compact` or `canonicalize`, which rebuilds the source to only include visible text (no markup), and then get the source.

```rust
fn iterate_on_spans(styled: &cursive::utils::markup::StyledString) -> String {
    styled.spans().map(|span| span.content).collect()
}

fn canonicalize(styled: &mut cursive::utils::markup::StyledString) -> String {
    styled.canonicalize();
    styled.source().to_string()
}

fn main() {
    let mut styled = cursive::utils::markup::cursup::parse("/blue{This} /white{is} /red{sparta}!");

    println!("`{}`", iterate_on_spans(&styled));
    println!("`{}`", canonicalize(&mut styled));
}
```

**Comment by xairaven** [comment]
Thank you so much again for your response! Everything worked out, and it's all functioning perfectly. I really enjoyed working with Cursive; it's a very convenient library. I’ve finished my first project with it - [xArpChat](https://github.com/xairaven/arpchat-rs).

Additionally, I completed the pull request with the example of colored text (#801). I gathered all your answers and included links to the documentation for the methods used.

I'm closing the issue. Thank you so much again for your help! If you find my [pull request](https://github.com/gyscos/cursive/pull/801) appropriate, I would be very happy to become a contributor to Cursive. Also, I apologize if any of my words sounded rude, I don't know English well.
